File: python_fastapi_release/app/pkg/banner/usecase.py
import abc
from typing import List, Optional, Union

from app.models import app_models


class BannerUseCase(abc.ABC):
    @abc.abstractmethod
    def get_user_banner(
        self, tag_ids: List[int], feature_id: int
    ) -> Union[app_models.Banner, None]:
        pass

    @abc.abstractmethod
    def get_banners(
        self,
        tag_id: int,
        feature_id: int,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> List[app_models.Banner]:
        pass

    @abc.abstractmethod
    def create_banner(
        self,
        banner: app_models.Banner,
        user: app_models.User,
    ) -> None:
        pass

    @abc.abstractmethod
    def delete_banner(self, banner_id: int) -> None:
        pass

    @abc.abstractmethod
    def update_banner(self, banner: app_models.Banner) -> None:
        pass

File: go_release/pkg/banner/usecase/usecase.go
package usecase

import (
	"context"

	"github.com/algakz/banner_service/models"
	"github.com/algakz/banner_service/pkg/banner"
)
type BannerUseCase struct {
	bannerRepo banner.Repository
}

func NewBannerUseCase(bannerRepo banner.Repository) *BannerUseCase {
	return &BannerUseCase{
		bannerRepo: bannerRepo,
	}
}

func (b *BannerUseCase) GetUserBanner(
	ctx context.Context,
	tag_ids []int,
	feature_id int,
) (models.Banner, error) {
	banner, err := b.bannerRepo.GetUserBanner(ctx, tag_ids, feature_id)
	return banner, err
}

func (b *BannerUseCase) GetBanners(
	ctx context.Context,
	tag_id int,
	feature_id int,
	limit int,
	offset int,
) ([]models.Banner, error) {
	banners, err := b.bannerRepo.GetBanners(ctx, tag_id, feature_id, limit, offset)
	return banners, err
}

func (b *BannerUseCase) CreateBanner(
	ctx context.Context,
	banner *models.Banner,
	user *models.User,
) (int, error) {
	banner_id, err := b.bannerRepo.CreateBanner(ctx, banner, user)
	if err != nil {
		return 0, err
	}
	return banner_id, nil
}

func (b *BannerUseCase) UpdateBanner(ctx context.Context, banner *models.Banner) error {
	err := b.bannerRepo.UpdateBanner(ctx, banner)
	return err
}

func (b *BannerUseCase) DeleteBanner(ctx context.Context, banner_id int) error {
	err := b.bannerRepo.DeleteBanner(ctx, banner_id)
	if err != nil {
		return err
	}
	return nil
}
